File: Cpp/0322_coin_change.cpp
// https://leetcode.cn/problems/coin-change/

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount+1, 0);
        for (int i = 1; i <= amount; ++i) {
            int tmp = amount + 1;
            for (int coin: coins) {
                int rem = i - coin;
                if (rem == 0) {
                    tmp = 0;
                }
                else if ((rem > 0) && (dp[rem] > 0)) {
                    tmp = min(tmp, dp[rem]);
                }
            }
            if (tmp == amount + 1) {
                dp[i] = -1;
            }
            else {
                dp[i] = tmp + 1;
            }
        }
        return dp[amount];
    }
};

File: Python3/0322_coin_change.py
# https://leetcode.cn/problems/coin-change/

class Solution:
    def coinChange(self, coins, amount):
        """
        Inputs:
            coins: list[int]
            amount: int
        Outputs:
            res: int
        """

        dp = [amount+1 for _ in range(amount+1)]
        dp[0] = 0
        for i in range(1, amount+1):
            for coin in coins:
                if coin <= i and dp[i-coin] + 1 < dp[i]:
                    dp[i] = dp[i-coin] + 1
        return dp[amount] if dp[amount] != amount+1 else -1